src/listen/ldap_sync/proto_ldap_sync_ldap.c

```diff
@@ -0,0 +1,1325 @@
+									    proto_ldap_sync_ldap_t);
+	fr_ldap_connection_t		*ldap_conn = talloc_get_type_abort(conn->h, fr_ldap_connection_t);
+
+	proto_ldap_dir_ctx		*dir_ctx;
+
+	if (ldap_conn->fd < 0) {
+	connection_failed:
+		if (fr_event_timer_in(thread, thread->el, &thread->conn_retry_ev,
+				      inst->handle_config.reconnection_delay,
+				      proto_ldap_connection_init, listen) < 0) {
+			FATAL("Failed inserting event: %s", fr_strerror());
+		}
+		return;
+	}
+
+	thread->li->fd = ldap_conn->fd;
+
+	MEM(dir_ctx = talloc_zero(inst, proto_ldap_dir_ctx));
+	if (!dir_ctx) goto connection_failed;
+
+	dir_ctx->main_listen = listen;
+	dir_ctx->conn = conn;
+	dir_ctx->child_listen = thread->li;
+
+	/*
+	 *	Set the callback which will handle the results of this query
+	 */
+	if (fr_event_fd_insert(conn, conn->el, ldap_conn->fd,
+			       _proto_ldap_socket_open_read,
+			       NULL,
+			       _proto_ldap_socket_open_error,
+			       dir_ctx) < 0) {
+					goto connection_failed;
+	}
+
+	/*
+	 *	Allocate the directory structure and send the query
+	 */
+	dir_ctx->msgid = fr_ldap_conn_directory_alloc_async(ldap_conn);
+
+	if (dir_ctx->msgid < 0) {
+		talloc_free(dir_ctx);
+		goto connection_failed;
+	}
+
+	/*
+	 *	Add a watch to catch closed LDAP connections
+	 */
+	fr_connection_add_watch_post(thread->conn, FR_CONNECTION_STATE_CLOSED,
+				     _proto_ldap_socket_closed, true, listen);
+}
+
+/** Callback triggered when parent listener app_io has its event list set
+ *
+ * Initiates the actual outbound LDAP connection
+ *
+ * @param[in] li	The parent listener.
+ * @param[in] el	Event list for this listener.
+ * @param[in] nr	Network handler.
+ */
+static void mod_event_list_set(fr_listen_t *li, fr_event_list_t *el, void *nr)
+{
+	proto_ldap_sync_ldap_t const	*inst = talloc_get_type_abort_const(li->app_io_instance, proto_ldap_sync_ldap_t);
+	proto_ldap_sync_ldap_thread_t	*thread = talloc_get_type_abort(li->thread_instance, proto_ldap_sync_ldap_thread_t);
+
+	/*
+	 *	Set up thread data
+	 */
+	thread->name = inst->handle_config.name;
+	thread->parent = li;
+	thread->el = el;
+	thread->nr = nr;
+	thread->inst = inst;
+
+	/*
+	 *	Initialise the connection
+	 */
+	proto_ldap_connection_init(el, fr_event_list_time(el), li);
+}
+
+static int mod_instantiate(module_inst_ctx_t const *mctx)
+{
+	proto_ldap_sync_ldap_t	*inst = talloc_get_type_abort(mctx->inst->data, proto_ldap_sync_ldap_t);
+	CONF_SECTION		*conf = mctx->inst->conf;
+	CONF_SECTION		*sync_cs;
+	sync_config_t		*sync_conf;
+	char const		*server;
+	char			*url;
+	size_t			i, len;
+	CONF_SECTION		*map_cs;
+	map_t const		*map;
+	tmpl_rules_t		parse_rules = {
+		.attr = {
+			.allow_foreign = true	/* Because we don't know where we'll be called */
+		}
+	};
+
+	/*
+	 *	Verify that the LDAP server configuration is valid, either
+	 *	distinct server and port or an LDAP url.
+	 */
+	fr_assert(inst->handle_config.server_str[0]);
+
+	server = inst->handle_config.server_str[0];
+
+	if (ldap_is_ldap_url(server)) {
```

Comment:
All this URL check code should be moved to a common library function and removed from rlm_ldap too